lightning/src/ln/channelmanager.rs

```diff
@@ -677,6 +667,30 @@ pub type SimpleRefChannelManager<'a, 'b, 'c, 'd, 'e, M, T, F, L> = ChannelManage
 /// essentially you should default to using a SimpleRefChannelManager, and use a
 /// SimpleArcChannelManager when you require a ChannelManager with a static lifetime, such as when
 /// you're using lightning-net-tokio.
+//
+// Lock order:
+// The tree structure below illustrates the lock order requirements for the different locks of the
+// `ChannelManager`. If two or more locks on the same branch is held at the same time, locks should
+// be taken in the order of the lowest to highest level in the tree.
+//
+// Lock order tree:
+// │
+// │──`channel_state`
+// │   │
+// │   │──`id_to_peer`
+// │   │
+// │   └──`per_peer_state`
+// │       │
+// │       │──`best_block`
```

Comment:
Ah yes you're absolutely correct.. Thanks for the correction, and sorry for the errors with the lock orders! I confusingly thought yesterday that only write locks created lock orders, and not read locks as well.

I've updated the tree with `best_block` after `pending_outbound_payments`, since I've confirmed that it's the case. Also added the `outbound_scid_aliases` lock after the `per_peer_state` in the tree, after confirming that lock order. Finally, I fixed some grammar mistakes + added the `total_consistency_lock` and `forward_htlcs` for the illustration to  be more correct. [aae7b4e](https://github.com/lightningdevkit/rust-lightning/pull/1656/commits/aae7b4eedf3d3a7c39dd023a396dd36133798db3)

In case it's helpful, I've created a simple lock order test for the `ChannelManager` that I think touches most of the lock orders. I'm yet to confirm that it's the case, so it would need more work if we'd want something similar:
```
#[test]
fn lock_order_test() {
let chanmon_cfgs = create_chanmon_cfgs(3);
let node_cfgs = create_node_cfgs(3, &chanmon_cfgs);
let node_chanmgrs = create_node_chanmgrs(3, &node_cfgs, &[None, None, None]);
let nodes = create_network(3, &node_cfgs, &node_chanmgrs);
let chan_1 = create_announced_chan_between_nodes(&nodes, 0, 1, channelmanager::provided_init_features(), channelmanager::provided_init_features());
let chan_2 = create_announced_chan_between_nodes(&nodes, 1, 2, channelmanager::provided_init_features(), channelmanager::provided_init_features());
send_payment(&nodes[0], &vec!(&nodes[1], &nodes[2])[..], 100000);
let _closed_channel_1 = close_channel(&nodes[0], &nodes[1], &chan_1.2, chan_1.3, true);
let _closed_channel_2 = close_channel(&nodes[1], &nodes[2], &chan_2.2, chan_2.3, true);
{
	let _tc_lock = nodes[0].node.total_consistency_lock.write().unwrap();
	let _channel_state_lock = nodes[0].node.channel_state.lock().unwrap();
	let _id_to_peer = nodes[0].node.id_to_peer.lock().unwrap();
	//let _short_to_chan_info = nodes[0].node.short_to_chan_info.write().unwrap();
	let _per_peer_state = nodes[0].node.per_peer_state.write().unwrap();
	let _outbound_scid_aliases = nodes[0].node.outbound_scid_aliases.lock().unwrap();
	let _pending_inbound_payments = nodes[0].node.pending_inbound_payments.lock().unwrap();
	let _pending_outbound_payments = nodes[0].node.pending_outbound_payments.lock().unwrap();
	let _best_block = nodes[0].node.best_block.write().unwrap();
	let _pending_events = nodes[0].node.pending_events.lock().unwrap();
	let _pending_background_events = nodes[0].node.pending_background_events.lock().unwrap();
	let _forward_htlcs = nodes[0].node.forward_htlcs.lock().unwrap();
}
nodes[0].node.get_and_clear_pending_events();
nodes[1].node.get_and_clear_pending_events();
nodes[2].node.get_and_clear_pending_events();
} 
```